enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,461 @@
+
+### Risks and Mitigations
+
+Not forking means that time sensitive fixes (like security) will need to be vendor-patched.
+
+Changes we propose might get rejected from the upstream. In such case we can
+vendor-patch or create fork.
+
+### Drawbacks
+
+MicroShift team will need to manage the go.mod importing with all the
+consequences of the Go modules versioning.
+
+## Alternatives (Not Implemented)
+
+### Forking
+
+Generic Device Plugin fork would provide us a place to quickly patch time
+sensitive issue (e.g. security, CVEs) however it would permanently add a
+recurring maintenance effort for MicroShift team.
+
+Trying to avoid that, team decided to skip forking and try upstream first approach,
+and only if it's really necessary fork the project.
+
+### Container approach
+
+An alternative to importing the code directly is the "container" approach.
+It involves publishing the forked Generic Device Plugin as an container image
+along with a manifest for MicroShift.
+This might align better with Kubernetes principles (containerization) and could
+offer broader reuse (but not necessarily by OpenShift right now), it presents
+several drawbacks:
+- Increased maintenance effort on MicroShift team:
+  As GDP would likely not be a part of the OpenShift payload and MicroShift team
+  gets help from ART in managing the RPMs. The MicroShift team does not have
+  experience in productizing images.
+- Size of the Generic Device Plugin image would likely be significantly larger
+  than increase in the MicroShift binary size from code import. This is
+  important consideration due to MicroShift's target environment.
+
+### Other device plugin solution
+
+Investigating this feature only three solutions were found:
+- Generic Device Plugin (discussed in this enhancement)
+- [Smarter Device Plugin](https://gitlab.com/arm-research/smarter/smarter-device-manager) (abandoned, last commit in Sept 2022)
+- [Akri](https://docs.akri.sh/)
+
+The Smarter Device Plugin was not considered as it is no longer actively maintained.
+
+Akri, while an interesting approach that abstracts the Device Plugin framework,
+was deemed too complex for our requirements. The MicroShift team concluded that
+integrating Akri would significantly increase maintenance effort across
+engineering, QE, and documentation.
+
+Since the Generic Device Plugin (GDP) can be disabled via configuration, users
+retain the flexibility to utilize upstream Akri if their needs require it.
+
+## Open Questions [optional]
+
+N/A
+
+## Test Plan
+
+Usually device plugins are tested using real hardware to make sure everything works.
+
+We'll need to get creative and implement some kind of dummy serial device to
+test communication between Pod utilizing the device and fake device
+(in the form of a daemon on the system).
+
+## Graduation Criteria
+
+Generic Device Plugin for MicroShift could start as GA.
+
+However, if we conclude that we're uncertain about "code import" method, we
```

Comment:
I think Technology preview is a better path since there's an uncertainty. Perhaps Support and TAM have an opinion on this as well.

Reply:
+1, I agree